src/main/java/com/conveyal/datatools/manager/controllers/api/ServerController.java

```diff
@@ -522,17 +639,44 @@ private static LoadBalancer getLoadBalancerForTargetGroup (EC2Info ec2Info, AWSS
      * Validate that ELB target group exists and is not empty and return associated load balancer for validating related
      * fields.
      */
-    private static LoadBalancer validateTargetGroupAndGetLoadBalancer(EC2Info ec2Info, Request req, AWSStaticCredentialsProvider credentials) {
+    private static EC2ValidationResult validateTargetGrouptLoadBalancerSubnetIdAndSecurityGroup(
+        EC2Info ec2Info,
+        AWSStaticCredentialsProvider credentials,
+        AmazonEC2 ec2Client
+    ) throws ExecutionException, InterruptedException {
+        EC2ValidationResult result = new EC2ValidationResult();
         if (isEmpty(ec2Info.targetGroupArn)) {
-            logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, "Invalid value for Target Group ARN.");
+            result.setInvalid("Invalid value for Target Group ARN.");
+            return result;
         }
         // Get load balancer for target group. This essentially checks that the target group exists and is assigned
         // to a load balancer.
         LoadBalancer loadBalancer = getLoadBalancerForTargetGroup(ec2Info, credentials);
         if (loadBalancer == null) {
-            logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, "Invalid value for Target Group ARN. Could not locate Target Group or Load Balancer.");
+            result.setInvalid("Invalid value for Target Group ARN. Could not locate Target Group or Load Balancer.");
+            return result;
+        }
+
+        // asynchronously execute the two validation tasks that depend on the load balancer info
+        List<Callable<EC2ValidationResult>> loadBalancerValidationTasks = new ArrayList<>();
+        loadBalancerValidationTasks.add(() -> validateSubnetId(loadBalancer, ec2Info, ec2Client));
+        loadBalancerValidationTasks.add(() -> validateSecurityGroupId(loadBalancer, ec2Info));
+        ExecutorService pool = Executors.newFixedThreadPool(loadBalancerValidationTasks.size());
+        for (Future<EC2ValidationResult> ec2ValidationResultFuture : pool.invokeAll(loadBalancerValidationTasks)) {
+            EC2ValidationResult loadBalancerValidationTaskResult = ec2ValidationResultFuture.get();
```

Comment:
This code seems very similar to the above code in validateEC2Config. Can it be DRYed out?

Reply:
They can't really be returned since they depend on first obtaining the load balancer info.